Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc

```diff
@@ -0,0 +1,177 @@
+    buffer.pop_front();
+    bufferNew.pop_front();
+
+    const std::string name = pD->name();
+
+    // Is it sensor? If yes, apply full sensor alignments
+    if (name == DDD_TOTEM_RP_SENSOR_NAME || name == DDD_CTPPS_DIAMONDS_SEGMENT_NAME ||
+        name == DDD_CTPPS_UFSD_SEGMENT_NAME || name == DDD_CTPPS_PIXELS_SENSOR_NAME ||
+        std::regex_match(name, std::regex(DDD_TOTEM_TIMING_SENSOR_TMPL))) {
+      unsigned int plId = pD->geographicalID();
+
+      if (alignments) {
+        const auto& ac = alignments->getFullSensorCorrection(plId);
+        pD->applyAlignment(ac);
+      }
+    }
+
+    // Is it RP box? If yes, apply RP alignments
+    if (name == DDD_TOTEM_RP_RP_NAME || name == DDD_CTPPS_DIAMONDS_RP_NAME || name == DDD_CTPPS_PIXELS_RP_NAME ||
+        name == DDD_TOTEM_TIMING_RP_NAME) {
+      unsigned int rpId = pD->geographicalID();
+
+      if (alignments) {
+        const auto& ac = alignments->getRPCorrection(rpId);
+        pD->applyAlignment(ac);
+      }
+    }
+
+    // create and add children
+    for (unsigned int i = 0; i < sD->components().size(); i++) {
+      const DetGeomDesc* sDC = sD->components()[i];
+      buffer.emplace_back(sDC);
+
+      // create new node with the same information as in sDC and add it as a child of pD
+      DetGeomDesc* cD = new DetGeomDesc(*sDC);
+      pD->addComponent(cD);
+
+      bufferNew.emplace_back(cD);
+    }
+  }
+}
+
+std::unique_ptr<DetGeomDesc> PPSGeometryESProducer::produceIdealGD(const IdealGeometryRecord& iRecord) {
+  // Get the DDDetector from EventSetup
+  auto const& det = iRecord.get(detectorToken_);
+
+  // Get the DDCompactView
+  cms::DDCompactView myCompactView(det);
+
+  // Build geo from compact view.
+  return DetGeomDescBuilder::buildDetGeomDescFromCompactView(myCompactView);
+}
+
+template <typename REC>
+std::unique_ptr<DetGeomDesc> PPSGeometryESProducer::produceGD(IdealGeometryRecord const& iIdealRec,
+                                                              std::optional<REC> const& iAlignRec,
+                                                              GDTokens<REC> const& iTokens,
+                                                              const char* name) {
+  // get the input GeometricalDet
+  auto const& idealGD = iIdealRec.get(iTokens.idealGDToken_);
+
+  // load alignments
+  edm::ESHandle<CTPPSRPAlignmentCorrectionsData> alignments;
+  if (iAlignRec) {
+    alignments = iAlignRec->getHandle(iTokens.alignmentToken_);
+  }
+
+  if (alignments.isValid()) {
+    if (verbosity_)
+      edm::LogVerbatim(name) << ">> " << name << " > Real geometry: " << alignments->getRPMap().size() << " RP and "
+                             << alignments->getSensorMap().size() << " sensor alignments applied.";
+  } else {
+    if (verbosity_)
+      edm::LogVerbatim(name) << ">> " << name << " > Real geometry: No alignments applied.";
+  }
+
+  DetGeomDesc* newGD = nullptr;
+  applyAlignments(idealGD, alignments.product(), newGD);
```

Comment:
`alignments.product()` will throw an exception if the `ESHandle` has no product. Guessing from the logic in `applyAlignments()` that does not seem to be the intention. If that is indeed the case, you could do something like
```cpp
CTPPSRPAlignmentCorrectionsData const* alignments = nullptr;
if (iAlignRec) {
  auto alignmentsHandle = iAlignRec->getHandle(iTokens.alignmentToken_);
  if (alignmentsHandle.isValid()) {
    alignments = alignmentsHandle.product();
  }
}
...
if (verbosity_) {
  if (alignments) {
    ...
  } else {
    ...
  }
}
```